vertical-pod-autoscaler/enhancements/4016-in-place-updates-support/README.md

```diff
@@ -0,0 +1,109 @@
+
+### Goals
+
+* Allow VPA to actuate without disruption,
+* Allow VPA to actuate more frequently, when it can actuate without disruption,
+* Allow VPA to actuate in situations where actuation by eviction doesn't work.
+
+### Non-Goals
+
+* Improved handling of injected sidecars
+  * Separate AEP will improve VPAs handling of injected sidecars.
+
+## Proposal
+
+Add new supported values of
+[`UpdateMode`](https://github.com/kubernetes/autoscaler/blob/71b489f5aec3899157b37472cdf36a1de223d011/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1/types.go#L124):
+
+* `InPlaceOnly` and
+* `InPlaceOrRecreate`.
+
+## Context
+
+[In-place update of pod resources KEP](https://github.com/kubernetes/enhancements/issues/1287) is available in alpha in
+[Kubernetes 1.27](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.27.md#api-change-3). The
+feature allows changing container resources while the container is running. It also adds 
+[`ResizePolicy`](https://github.com/kubernetes/api/blob/8360d82aecbc72aa039281a394ebed2eaf0c0ccc/core/v1/types.go#L2448)
+field to Container. This field indicates that the container should be restarted when its resources
+
+## Design Details
+
+VPA Updater will execute in-place updates or eviction-updates in the newly added `InPlaceOnly` and `InPlaceOrRecreate`
+modes. 
+
+### Applying Updates During Pod Admission
+
+For VPAs in `InPlaceOnly` and `InPlaceOrRecreate` modes VPA Admission Controller will apply updates to starting pods,
+like it does for VPAs in `Initial`, `Auto`, and `Recreate` modes.
+
+### Applying Disruption-free Updates
+
+When update changes only resources for which container indicates that it doesn't require a restart then VPA can attempt
+to actuate the change without disrupting the pod. VPA Updater will:
+* attempt to actuate such updates in place,
+* attempt to actuate them if difference between recommendation and request is at least 10%
+  * even if pod has been running for less than 24h,
+* if necessary execute only partial updates (if some changes would require a restart).
+
+### Applying Disruptive Updates
+
+In both `InPlaceOnly` and `InPlaceOrRecreate` modes VPA updater will attempt to apply updates that require container
+restart in place. It will update them under conditions that would trigger update with `Recreate` mode. That is it will
+apply disruptive updates if:
+
+* Sum of pods requests is below sum of recommendations `LowerBound` or
+* Sum of pods requests is above sum of recommendations `UppelBound` or
+* Difference between sum of pods requests and sum of recommendation `Target`s is more than 10% and the pod has been
+  running undisrupted for at least 24h.
+  * Successful disruptive update counts as disruption here (and prevents further disruptive updates to the pod for 24h).
+
+In `InPlaceOrRecreate` mode (but not in `InPlaceOnly` mode) VPA updater will evict pod to actuate a recommendation if it
+attempted to apply the recommendation in place and failed.
+
```

Comment:
I'd like to have the order of updater actions explicitly stated.
I expected it would attempt in-place updates first but a) it's only my assumption b) I'm not sure it always would be the case.
I can see at least the following possibilities:
* all resources can be updated in-place and the change matches eviction criteria: attempt in-place, evict on error?
* some resources can be updated in-place, some not and the change matches eviction criteria: would VPA evict already, w/o attempting in-place update (this seems contradictory to "if necessary execute only partial updates" earlier), does this depend on which resource matches eviction criteria, other?
* all or some resources can be updated in-place but the change doesn't match eviction criteria: in-place only?
* no possibility for in-place update: go to the existing updater flow.

We also have priority ordering in updater, maybe it's worth adding a sentence how it would be changed?

Reply:
I'm not sure I understand what you mean by "order of updater actions". Do you mean order of operations in updaters [`RunOnce`](https://github.com/kubernetes/autoscaler/blob/114a35961a85efdf3f36859350764e5e2c0c7013/vertical-pod-autoscaler/pkg/updater/logic/updater.go#L109)?

Or do you mean since now there are up to 3 different things we could attempt for any given pod, in which of them we actually do?

The things we could do:

1. Apply full update by eviction,
2. Apply full update in-place, trigger some container restarts,
3. Apply partial update in-place, don't restart any containers

We definitely need to try 2. before attempting 1. Attempt 1. only if 2. failed.

If both 2. and 3. are an option I think we should do 2. (there's some reason we didn't do the smaller disruption-free update earlier so let's move quickly).

I'm not updating AEP yet because I'm not sure what to put there.

Reply:
2. for `InPlaceOrRecreate` and 3. for `InPlaceOnly`?

Reply:
2. for both `InPlaceOrRecreate` and `InPlaceOnly`